docs/travis.md

```diff
@@ -0,0 +1,102 @@
+
+The main Travis configuration file is `.travis.yml`.  The specific
+test suites we have are listed in the `matrix` section in that file.
+We've configured it to use a few helper scripts for each job:
+
+* `tools/travis/setup-$TEST_SUITE`: A Zulip script that sets up the
+  test environment (E.g. installing dependencies).
+  * For the backend and frontend suites, this is a thin wrapper around
+    `tools/provision`, aka the development environment provision script.
+  * For the production suite, this is a more complicated process
+    because of all the packages Travis installs.  See the comments in
+    `tools/travis/setup-production` for details.
+* `tools/travis/$TEST_SUITE`: A zulip script containing the actual
+  code for the test.
+
+The main purpose of the distinction between the two is that if the
+`setup-backend` job fails, Travis CI will report it as a build
+infrastructure failure, whereas if the `backend` job fails, it'll be
+reported as a test failure.
+
+An important detail is that Travis CI hides most build phases other
+than the actual test; you can see this easily by looking at the line
+numbers in the Travis CI output.  There are actually a bunch of phases
+(e.g. the project's setup job, downloading caches near the beginning,
+uploading caches at the end, etc.), and if you're debugging our
+configuration, you'll want to look at these closely.
+
+## Useful debugging tips and tools
+
+* For performance issues,
+  [this travis stats tool](https://scribu.github.io/travis-stats/#zulip/zulip/master)
+  can give you test runtime history data that can help with
+  determining when a performance issue was introduced and whether it
+  was fixed.
+
+* You can [sign up your personal repo for Travis CI][travis-fork] so
+  that every remote branch you push will be tested, which can be
+  helpful when debugging something complicated.
+
+[travis-fork]: git-guide.html#step-3-configure-travis-ci-continuous-integration
+
+## Performance optimizations
+
+### Caching
+
+An important element of making Travis CI perform effectively is
+caching the provisioning of a Zulip development environment.  In
+particular, we cache the following across builds:
+
+* Python virtualenvs
+* node_modules directories
+* Built/downloaded emoji sprite sheets and data
+
+This has a huge impact on the performance of running tests in Travis
+CI; without these caches, the average test time would be several times
+longer.
+
+We have designed these caches carefully (they are also used in
+production and the Zulip development environment) to ensure that each
+is named by a hash of its dependencies, so Zulip should always be
+using the same version of dependencies it would have used had the
+cache not existed.  In practice, bugs are always be possible, so be
+mindful of this possibility.
+
+A consequence of this caching is that test jobs for branches which
+modify `package.json`, `requirements/`, and other key dependencies
+will be significantly slower than normal.
```

Comment:
I'd spell this out a bit more: "than normal, because they won't get to benefit from the cache."